sdlib/d/gc/scanner.d

```diff
@@ -137,23 +138,30 @@ private:
 		 * and we shoudl stop.
 		 */
 		static hasWork(Scanner* w) {
-			return w.cursor != 0 || w.activeThreads == 0;
+			return w.workItems != 0 || w.activeThreads == 0;
 		}
 
 		auto w = (cast(Scanner*) &this);
 		mutex.waitFor(w.hasWork);
 
-		if (w.cursor == 0) {
+		if (w.workItems == 0) {
 			return false;
 		}
 
 		activeThreads++;
 
-		auto stop = w.cursor;
-		auto start = stop - 1;
+		if (--w.workItems == 0) {
+			worker.refill(w.worklist[w.cursor .. w.cursor + 1]);
+			if (w.cursor > 0) {
+				--w.cursor;
+				w.workItems = getWorkItems(w.worklist[w.cursor]);
+			}
+
+			return true;
+		}
```

Comment:
There is added work here and on the insert.

It might be too much, I don't know. But it's only a handful of operations (2 comparisons + bit shift)